src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+
+
+# In this tutorial, we will learn:
+# 
+#   * How to apply the adjoint method for sensitivity analysis in Gridap
+#   * How to do topology optimization in Gridap
+# 
+# We recommend that you first read the [Electromagnetic scattering tutorial](https://gridap.github.io/Tutorials/dev/pages/t012_emscatter/#Tutorial-12:-Electromagnetic-scattering-in-2D-1) to make sure you understand the following points:
+# 
+#   * How to formulate the weak form for a 2d time-harmonic electromagnetic problem (a scalar Helmholtz equation)
+#   * How to implement a perfectly matched layer (PML) to absorb outgoing waves
+#   * How to impose periodic boundary conditions in Gridap
+#   * How to discretize PDEs with complex-valued solutions
+# 
+# ## Problem statement
+# 
+# We are optimizing the electric field at the center of a $r_d=100$ nm design region with a Hz-polarized plane wave and  smallest distance constraint at $r_s = 10$ nm. The computational cell is of height $H$ and length $L$, and we employ a perfectly matched layer (PML) thickness of $d_{pml}$ to implement outgoing (radiation) boundary conditions for this finite domain. 
+# ![](../assets/TopOptEMFocus/Illustration.png)
+# 
+# The goal is find the arrangment of the silver material in the gray region that maximize the electric field magnitude at the center point. Every "pixel" in the gray region would be a degree of freedom that can vary between silver (shown in black below) and air (shown in white below). This is called density-based [topology optimization (TO)](https://en.wikipedia.org/wiki/Topology_optimization). 
+# 
+# ## Formulation
+# 
+# From Maxwell's equations, considering a time-harmonic electromagnetic field polarized so that the electric field is in-plane and the magnetic field is out-of-plane (described by a scalar $H$ equal to the z-component), we can derive the governing equation of this problem in 2D (Helmholtz equation) [1]: 
+# 
+# ```math
+# \left[-\nabla\cdot\frac{1}{\varepsilon(x)}\nabla -k^2\mu(x)\right] H = f(x), 
+# ```
+# 
+# where $k=\omega/c$ is the wave number in free space and $f(x)$ is the source term (which corresponds to a magnetic current density in Maxwell's equations).
+# 
+# In order to simulate this scattering problem in a finite computation cell, we need outgoing (radiation) boundary conditions such that all waves at the boundary would not be reflected back since we are simulating an infinite space.  One commonly used technique to simulate such infinite space is through the so called "perfectly matched layers" (PML) [2]. Actually, PML is not a boundary condition but an artificial absorbing "layer" that absorbs waves with minimal reflections (going to zero as the resolution increases).  There are many formulations of PML. Here, we use one of the most flexible formulations, the "stretched-coordinate" formulation, which takes the following replace in the PDE [3]:
```

Comment:
```suggestion
# In order to simulate this scattering problem in a finite computational domain, we need outgoing (radiation) boundary conditions to prevent waves from reflecting back from the boundaries of the domain.  We employ the well-known technique of "perfectly matched layers" (PML) [2], which are an an artificial absorbing layer adjacent to the boundaries that absorbs waves with minimal reflections (going to zero as the resolution increases). The "stretched-coordinate" formulation of PML correspond to a simple transformation of the PDE [3]:
```